firedrake/bcs.py

```diff
@@ -95,6 +80,173 @@ def __init__(self, V, g, sub_domain, method="topological"):
 
+    def function_space(self):
+        '''The :class:`.FunctionSpace` on which this boundary condition should
+        be applied.'''
+
+        return self._function_space
+
+    @utils.cached_property
+    def domain_args(self):
+        r"""The sub_domain the BC applies to."""
+        # Define facet, edge, vertex using tuples:
+        # Ex:
+        #                                      key.. one = 1 to avoid 0 * sqrt(...)
+        # facet  = ((1, ), )       ->     (one + 1) * sqrt(1)
+        # edge   = ((1, 2), )      ->     (one + 1) * sqrt(1) + (one + 2) * sqrt(2)
+        # vertex = ((1, 2, 4), )   ->     (one + 1) * sqrt(1) + (one + 2) * sqrt(2) + (one + 4) * sqrt(3)
+        #
+        # Multiple facets: (1, 2, 4) := ((1, ), (2, ), (4, )) returns ((1, 2, 4), )
+        #
+        # One facet and two edges:
+        # ((1,), (1, 3), (1, 4)) returns ((2, 2 + 4 * sqrt(2), 2 + 5 * sqrt(2)), )
+        #
+        # TODO: better ways?
+        #
+        sub_d = self.sub_domain
+        if isinstance(sub_d, str):
+            return (sub_d, )
+        sub_d = as_tuple(sub_d)
+        sub_d = [as_tuple(sub_d[i]) for i in range(len(sub_d))]
+        sd = []
+        for i in range(len(sub_d)):
+            v = 0.0
+            for ii in range(len(sub_d[i])):
+                v += (1 + sub_d[i][ii]) * np.sqrt(ii + 1)
+            sd.append(v)
+        return (as_tuple(sd), )
+
+    @utils.cached_property
+    def nodes(self):
+        '''The list of nodes at which this boundary condition applies.'''
+        sub_d = self.sub_domain
+        if isinstance(sub_d, str):
+            bcnodes = self._function_space.boundary_nodes(sub_d, self.method)
+            if isinstance(self._function_space.finat_element, finat.Hermite) and \
+               self._function_space.mesh().topological_dimension() == 1:
+                bcnodes = bcnodes[::2]  # every second dof is the vertex value
+        else:
+            sub_d = as_tuple(sub_d)
+            sub_d = [as_tuple(sub_d[i]) for i in range(len(sub_d))]
+            bcnodes = np.empty(0, dtype=int)
+            for iboun in range(len(sub_d)):
+                _bcnodes = self._function_space.boundary_nodes(sub_d[iboun][0], self.method)
+                if isinstance(self._function_space.finat_element, finat.Hermite) and \
+                   self._function_space.mesh().topological_dimension() == 1:
+                    _bcnodes = _bcnodes[::2]  # every second dof is the vertex value
+                for i in range(1, len(sub_d[iboun])):
+                    # intersection of facets
+                    # Edge conditions have only been tested with Lagrange elements.
+                    # Need to expand the list.
+                    if not isinstance(self._function_space.finat_element, finat.Lagrange):
+                        raise TypeError("Currently, edge conditions have only been tested with Lagrange elements")
```

Comment:
This algorithm extracts the edge (or vertex) nodes for H^1 functions by taking intersections of facet nodes.
We need this to apply edge Dirichlet BCs to a facet EquationBC.
For instance if we have an EquationBC on subdomain 1 and a DirichletBC on subdomain 3, we need to apply an edge DirichletBC at the intersection of 1 and 3 when we assemble the facet EquationBC on subdomain 1.

If we want to apply edge EquationBCs for facet EquationBCs, we need something better as we would also need edge iteration sets, but I think for now this can deal with many problems.

Reply:
This part looks better now, too.